File: scala/pascals-triangle/src/main/scala/PascalsTriangle.scala

object PascalsTriangle {
  def rows(i: Int): List[List[Int]] =
    if (i <= 0) List() else
      (2 to i).foldLeft(List(List(1))) { (acc, _) =>
        acc ++ List(List(1) ++ acc.last.dropRight(1).zipWithIndex.map { n => n._1 + acc.last(n._2 + 1) } ++ List(1))
      }


}

File: kotlin/pascals-triangle/src/main/kotlin/PascalsTriangle.kt
object PascalsTriangle {

    fun computeTriangle(rows: Int): List<List<Int>> =
        if (rows <= 0) { listOf() } else
        (2..rows).fold(listOf(listOf(1))) { acc, _ ->
            acc + listOf( listOf(1) + acc.last().dropLast(1).withIndex()
                    .map { it.value + acc.last()[it.index + 1] } + listOf(1))
        }
}
